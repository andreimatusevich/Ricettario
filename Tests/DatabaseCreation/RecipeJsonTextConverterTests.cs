using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using ApprovalTests;
using ApprovalTests.Reporters;
using MoreLinq;
using Newtonsoft.Json;
using NUnit.Framework;
using Ricettario.Shared.Model;

namespace Tests.DatabaseCreation
{
    public static class Extensions
    {
        public static string TrimEx(this string text)
        {
            return text?.Trim(' ', ';', ',', ':');
        }

        public static bool IsAlike(this string text, params string[] ethalons)
        {
            var normalized = text?.TrimEx().ToLower();
            return ethalons.Select(e => e?.TrimEx().ToLower()).Any(e => e == normalized);
        }
    }

    public class TextRecipeParser
    {
        private readonly IFactory<Recipe> _factory;

        private static Dictionary<string, Action<Recipe, IEnumerable<string>>> _rules = new Dictionary
            <string, Action<Recipe, IEnumerable<string>>>
        {
            {
                "Name", (recepie, lines) => recepie.Name = lines.First().Trim()
            },
            {
                "RecipeYield",
                (recepie, lines) => recepie.RecipeYield = String.Join(" ", lines.Skip(1).TakeWhile(l => !String.IsNullOrEmpty(l)))
            },
            {
                "Ingredients",
                (recepie, lines) =>
                    recepie.RecipeIngredients =
                        lines.SkipWhile(l => !l.IsAlike("Ingredients"))
                            .Skip(1)
                            .SkipWhile(l => String.IsNullOrEmpty(l))
                            .TakeWhile(l => !String.IsNullOrEmpty(l)).ToList()
            },
            {
                "Instructions",
                (recepie, lines) =>
                    recepie.RecipeInstructions =
                        lines.SkipWhile(l => !l.IsAlike("Method", "Directions"))
                            .Skip(1).Where(l => !String.IsNullOrEmpty(l)).ToList()
            }
        };

        public TextRecipeParser(IFactory<Recipe> factory)
        {
            _factory = factory;
        }

        public Recipe FromLines(IEnumerable<string> lines)
        {
            var normalizedLines = lines.SkipWhile(l => String.IsNullOrEmpty(l.TrimEx())).Select(l => l.Trim()).ToList();
            var recepie = _factory.Create();
            _rules.Values.ForEach(r => r(recepie, normalizedLines));
            return recepie;
        }
    }

    public class RecipeJsonTextConverterTests
    {
        [Test]
        [UseReporter(typeof(WinMergeReporter))]
        public void Simple()
        {
            var fileName = @"..\..\DatabaseCreation\Data\simple_recepie.txt";
            TestParse(fileName);
        }

        [Test]
        [UseReporter(typeof(WinMergeReporter))]
        public void GoogleDocs()
        {
            var fileName = @"..\..\DatabaseCreation\Data\google_docs_recepie.txt";
            TestParse(fileName);
        }

        private static void TestParse(string fileName)
        {
            var lines = File.ReadAllLines(fileName);

            var parser = new TextRecipeParser(new TestRecipeFactory());
            var recepie = parser.FromLines(lines);

            var result = JsonConvert.SerializeObject(recepie, Formatting.Indented,
                new JsonSerializerSettings {NullValueHandling = NullValueHandling.Ignore});
            Approvals.Verify(result);
        }
    }


    public class TestRecipeFactory: IFactory<Recipe>
    {
        public Recipe Create()
        {
            return new Recipe()
            {
            };
        }
    }
}