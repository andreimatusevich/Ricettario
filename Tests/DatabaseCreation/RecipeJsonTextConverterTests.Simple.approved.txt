{
  "Version": "0001-01-01T00:00:00",
  "Name": "Braised Lamb Shanks with White Bean Purée",
  "RecipeIngredients": [
    "4 large or 8 small Australian lamb shanks",
    "2 tablespoons olive oil",
    "1 large onion, peeled and roughly chopped",
    "2 medium carrots, peeled and roughly diced",
    "2 celery sticks, washed and roughly chopped",
    "4 cloves garlic, crushed",
    "2 ½ pints lamb or beef stock",
    "3 bay leaves, dried",
    "1 pound dried white beans that have been soaked in water overnight",
    "14 ounce canned tomatoes"
  ],
  "RecipeInstructions": [
    "Seal the lamb shanks by pan frying in a dry pan, or the casserole dish itself, to lightly brown.",
    "In a large casserole, heat the oil and brown the onion, carrots, celery and garlic. Lay the shanks in the pan, pour in the stock and allow to come to a simmer. Add the bay leaves.",
    "Tip the beans into a muslin cloth and tie to secure. Add the beans to the shanks and cover.",
    "Cook in the oven at 350°F for 1 hour. Check to see that the beans are cooked. When the beans are done, remove them  from the shanks, along with 1 cup of the cooking liquid. Add the tomatoes to the shanks and continue cooking for up to 1 hour more or until the meat is tender and coming away from the bone.",
    "Puree the beans in a food processor with enough of the cooking liquid to make a paste; season with salt and white pepper and store in a warm place or in a microwave proof container.",
    "When the shanks are ready, check seasoning and remove the bay leaves; spoon some of the reheated or hot bean puree into the centre of the plates and put a shank/s on top. Ladle some of the cooking juices around and over the shanks and serve cooked green vegetables or a ratatouille."
  ],
  "RecipeYield": "Servings 4"
}
-------------------
Name:
	Braised Lamb Shanks with White Bean Purée

Yield:
	Servings 4

Ingredients:
	4 large or 8 small Australian lamb shanks
	2 tablespoons olive oil
	1 large onion, peeled and roughly chopped
	2 medium carrots, peeled and roughly diced
	2 celery sticks, washed and roughly chopped
	4 cloves garlic, crushed
	2 ½ pints lamb or beef stock
	3 bay leaves, dried
	1 pound dried white beans that have been soaked in water overnight
	14 ounce canned tomatoes

Directions:
	Seal the lamb shanks by pan frying in a dry pan, or the casserole dish itself, to lightly brown.
	In a large casserole, heat the oil and brown the onion, carrots, celery and garlic. Lay the shanks in the pan, pour in the stock and allow to come to a simmer. Add the bay leaves.
	Tip the beans into a muslin cloth and tie to secure. Add the beans to the shanks and cover.
	Cook in the oven at 350°F for 1 hour. Check to see that the beans are cooked. When the beans are done, remove them  from the shanks, along with 1 cup of the cooking liquid. Add the tomatoes to the shanks and continue cooking for up to 1 hour more or until the meat is tender and coming away from the bone.
	Puree the beans in a food processor with enough of the cooking liquid to make a paste; season with salt and white pepper and store in a warm place or in a microwave proof container.
	When the shanks are ready, check seasoning and remove the bay leaves; spoon some of the reheated or hot bean puree into the centre of the plates and put a shank/s on top. Ladle some of the cooking juices around and over the shanks and serve cooked green vegetables or a ratatouille.

Description:

Author:

Tags: